geometry/spline.cpp

```diff
@@ -60,20 +60,40 @@ void Spline::ReplacePoint(PointD const & pt)
   AddPoint(pt);
 }
 
-bool Spline::IsProlonging(PointD const & pt) const
+void Spline::AddOrProlongPoint(PointD const & pt, double minSegSqLength, bool checkAngle)
 {
-  size_t const sz = m_position.size();
-  if (sz < 2)
-    return false;
+  if (m_position.empty())
+  {
+    m_position.push_back(pt);
+    return;
+  }
 
   PointD dir = pt - m_position.back();
   if (dir.IsAlmostZero())
-    return true;
-  dir = dir.Normalize();
+    return;
+  double len = dir.SquaredLength();
+  ASSERT_GREATER(len, 0, ());
+
+  if (m_position.size() > 1 && len < minSegSqLength)
+  {
+    ReplacePoint(pt);
+    return;
+  }
+
+  len = sqrt(len);
+  dir = dir / len;
 
-  ASSERT(!m_direction.empty(), ());
   // Some cos(angle) == 1 (angle == 0) threshold.
-  return std::fabs(DotProduct(m_direction.back(), dir)) > 0.995;
+  if (checkAngle && !m_direction.empty() && std::fabs(DotProduct(m_direction.back(), dir)) > 0.995)
+  {
+    ReplacePoint(pt);
+  }
+  else
+  {
+    m_position.push_back(pt);
```

Comment:
Can the number of allocations be reduced by preliminary calling .reserve for these containers?